tsdb/shard.go

```diff
@@ -1959,48 +2096,197 @@ func (fs *MeasurementFieldSet) marshalMeasurementFieldSetNoLock() (marshalled []
 	}
 }
 
-func (fs *MeasurementFieldSet) load() error {
-	fs.mu.Lock()
-	defer fs.mu.Unlock()
+func marshalFieldChanges(changeSet ...FieldChanges) ([]byte, error) {
+	fcs := internal.FieldChangeSet{
+		Changes: nil,
+	}
+	for _, fc := range changeSet {
+		for _, f := range fc {
+			mfc := &internal.MeasurementFieldChange{
+				Measurement: f.Measurement,
+				Delete:      f.Delete,
+			}
+			if f.Field != nil {
+				mfc.Field = &internal.Field{
+					Name: []byte(f.Field.Name),
+					Type: int32(f.Field.Type),
+				}
+				fcs.Changes = append(fcs.Changes, mfc)
+			}
+		}
+	}
+	b, err := proto.Marshal(&fcs)
```

Comment:
`proto.Buffer.Marshal` would let you leave space for the size at the front, and then we only need one allocation here and (especially) only one `Write` system call later.

e.g.

```
buf := protoc.NewBuffer(nil)
buf.EncodeFixed64(0)
buf.Marshal(&fcs)
binary.LittleEndian.PutUint64(buf.bytes()[0:8], uint64(len(b)-8))
...
return buf.bytes()
```